### PR TITLE
Latest version of employment and income metric update/de-bug

### LINKED ISSUE #262: Income and employment ACS update/bug-fix
- [ ] Switch .R scripts to .qmd
- [ ] Add summary statistics and visualizations to Quarto docs to visualize calculated metrics
- [ ] Change missing value codes (e.g. `9999999`)  to `NA` immediately after dropping labels and before any analysis
- [ ] Add automated tests to compare independent cities/counties in both data sets
- [ ] Make a function for the IPUMS API pull

### PR DESCRIPTION
**Pull request by jwalsh28**
Closes #262 

Notes:
- Includes a function for the IPUMS API extract (included in a new folder titled functions).
- Added a text file to the raw folder in data/temp because I could not get Git to create the folder while it was emtpy. This is where the API function stores the IPUMS extracts.
- Consolidated the all and subgroup files for county and place.

For @jameson-c 
Please review the following files:
income_metrics_county.qmd
employment_metrics_place.qmd
historical_comparison_income_metric.qmd

Note that these files also rely on a custom function which is titled extract_ipums.R

For @awunderground please review:
income_metrics_place.qmd
employment_metrics_county.qmd
historical_comparison_employment_metric_county.qmd
extract_ipums.R

### CODE REVIEW

File: 09_employment/employment_metrics_place.qmd
Comment (awunderground) [inline 09_employment/employment_metrics_place.qmd:384]: Does this make sense vs/ published statistics?

File: 09_employment/employment_metrics_place.qmd
Comment (awunderground) [inline 09_employment/employment_metrics_place.qmd:549]: Can you look at this for `"All"` and subgroups separately? Repeat for all metrics.

File: 09_employment/employment_metrics_place.qmd
Comment (awunderground) [inline 09_employment/employment_metrics_place.qmd:466]: This `size_flag` does not make sense. The quality flag needs to be redone.

File: 09_employment/employment_metrics_county.qmd
Comment (awunderground) [inline 09_employment/employment_metrics_county.qmd:240]: Same recommendation as in the income script.

File: 09_employment/employment_metrics_county.qmd
Comment (awunderground) [inline 09_employment/employment_metrics_county.qmd:457]: This size_flag does not make sense. The quality flag needs to be redone.

File: 01_financial-well-being/income_metrics_place.qmd
Comment (awunderground) [inline 01_financial-well-being/income_metrics_place.qmd:57]: Fix description.

File: 01_financial-well-being/income_metrics_place.qmd
Comment (awunderground) [inline 01_financial-well-being/income_metrics_place.qmd:84]: Should the `size_flag` be weighted by population?

File: 01_financial-well-being/income_metrics_place.qmd
Comment (awunderground) [inline 01_financial-well-being/income_metrics_place.qmd:421]: Need to update data quality.

**Review by awunderground — CHANGES_REQUESTED** [review]
I think the estimates are correct. We can improve the clarity of the code. We need to improve the data quality flags and potentially include suppression. 

I found some funky results for the county data. Some of the metrics have are supported by one or two observations since we don't use suppression. We should. 

We also don't mark anything `3` because of small sample size. Note the following comment from Kevin's code:

```sas
/* this creates the quality metric. Per discussion with Greg, 
only ACS metrics with sample size < 100 should be be marked as "3"
This code gives a county a "1" if they have a large sample size and
at least 75% of the data comes from county. If the county has a large
sample size and less than 75% of the data comes from the county, the 
county gets a "2." Small sample size gives a "3" */

```

File: data/temp/raw/Document1.txt
Comment (jwalsh28) [reply data/temp/raw/Document1.txt:1]: I was just having trouble adding the new "raw" folder while it was empty (I think becuase of the GitIgnore). I will delete on this round of edits and hopefully the folder stays.

File: functions/API/extract_ipums.R
Comment (jwalsh28) [reply functions/API/extract_ipums.R:56]: The way I have this set up is that as soon as the extract is read in it is renamed and includes "umf" in the title - the function looks for "umf" in the title of files in the temp/raw folder and will ignore these. That will be the standard for all programs with IPUMS extracts, such that if all code imediately renames the download to include "umf" and we have logic ignoring these files they will not be affected.

Comment (awunderground) [reply functions/API/extract_ipums.R:56]: Good call. My bad.

File: data/temp/raw/Document1.txt
Comment (awunderground) [reply data/temp/raw/Document1.txt:1]: I typically do this: https://stackoverflow.com/questions/115983/how-do-i-add-an-empty-directory-to-a-git-repository

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jwalsh28) [reply 01_financial-well-being/income_metrics_county.qmd:57]: Good catch, because this will be the single pull for all the predictors using ACS micro data I'm considering a title like "Microdata pull for Mobility Metric Predictors. American Community Survey, years 2018, 2019 and 2022 (1 and 5-year)."

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jwalsh28) [reply 01_financial-well-being/income_metrics_county.qmd:165]: Agreed, will add.

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jameson-c) [inline 01_financial-well-being/income_metrics_county.qmd:46]: Although it isn't particularly problematic in this case, a more secure way of doing this would be to define the API key in an environment that is blocked from merging by .gitignore. Not imperative but I figured I would mentioned. API keys generally shouldn't be pushed to git, so it might be helpful to add a comment that says #DO NOT PUSH THIS AND YOU ONLY NEED TO RUN ONCE  https://cran.r-project.org/web/packages/httr/vignettes/secrets.html

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jameson-c) [inline 01_financial-well-being/income_metrics_county.qmd:56]: throws an error because "raw" folder does not exist. I add suggested changes in extract_ipums to create an empty folder if it is missing.

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jameson-c) [inline 01_financial-well-being/income_metrics_county.qmd:272]: throws Warning message:
In left_join(acs_clean, puma_county_crosswalk, by = c("crosswalk_period",  :
  Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 1 of `x` matches multiple rows in `y`.
ℹ Row 4751 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship = "many-to-many"` to silence this warning.

Comment (jameson-c) [reply 01_financial-well-being/income_metrics_county.qmd:272]: There is a person who has multiple matches in "crosswalk_period", "statefip","puma"... well more than one, it looked like the population is doubling. from 41 million to 85 million. Is this error present when you run this line?

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jameson-c) [inline 01_financial-well-being/income_metrics_county.qmd:171]: From the xml document: "Beginning in 2000, the race question changed substantially to allow respondents to report as many races as they felt necessary to describe themselves. In earlier years, only one race response was coded. Beginning in 2020, the Census Bureau updated the questionnaire text, processing, and coding of the race and Hispanic origin questions, resulting in major changes to the distribution of race and Hispanic origin categories. As a result, users should proceed with caution when comparing RACE and HISPAN in 2019-prior samples with 2020-onward samples. See the comparability tab for more details. " I think that this evaluates race on either side of this cutoff. Not sure where this 'comparability' tab is but might be worth checking out.

File: 01_financial-well-being/income_metrics_county.qmd
Comment (jameson-c) [reply 01_financial-well-being/income_metrics_county.qmd:119]: In chunk 18, I get the following error otherwise:
Error: cannot allocate vector of size 653.0 Mb
Error during wrapup: cannot allocate vector of size 326.5 Mb
Error: no more error handlers available (recursive errors?); invoking 'abort' restart

File: 01_financial-well-being/historical_comparison_income_metric_county.qmd
Comment (jameson-c) [inline 01_financial-well-being/historical_comparison_income_metric_county.qmd:47]: Just wondering whether the following discrepancies are acceptable- counties_old has 3100 counties compared with 3143 in counties_new. counties_new has more NA values as well. Quite similar distributions though.

**Review by jameson-c — CHANGES_REQUESTED** [review]
Great job! submitting a PR to this in a branch `jc_edits` for aaron's review with my edits.

- there were a few hard-coded numbers that I did away with in the suppression factor calculation
- there was some repetitive syntax which I corrected in the extract_ipums.r file, and an error when the raw directory did not already exist
- extract_ipums.r required some additional commenting and function documentation, which I added
- A number of the graphs weren't readable due to long labels. I added a wrapper to make the labels readable.
- I changed a couple of filter-based drop na arguments to use drop_na instead for efficiency reasons

**Review by jameson-c — APPROVED** [review]
Great work!

**Review by jameson-c — APPROVED** [review]

**Comment by jwalsh28** [comment]
@awunderground waiting on your approval before we can merge

**Review by awunderground — APPROVED** [review]

**Pull request merged**
